### ISSUE TITLE
DataGrid fires CustomFilter 3 times x N rows at first time loading

### ISSUE THREAD

**Issue by barruka** [Megabit/Blazorise#3637]
**Describe the bug**
Using the example provided in the doc ```https://blazorise.com/docs/extensions/datagrid/features``` to enable a CustomFilter ```CustomFilter="@OnCustomFilter"``` I realize that CustomFilter function fires (3 x N rows) times, so in the example we have 5 rows so CustomFilter fires 15 times.

**To Reproduce**
1. Copy definition and code provided in 'Custom Filtering' section at 'https://blazorise.com/docs/extensions/datagrid/features'
2. Add a private class Employee to make it compile and work:
```
private class Employee
   {
       public string FirstName { get; set; } = String.Empty;
   }
```
3. Set a breakpoint inside ```private bool OnCustomFilter(Employee model)```
4. Debug & see how ```OnCustomFilter``` is called 15 times (3 x 5 rows)

**Expected behavior**
At first run, when DataGrid is loading, CustomFilter callback must not be called because the data is loading and the user is not able to input a filter at that moment.

**Comment by David-Moreira** [comment]
Hello @barruka 

So the way `CustomFilter` works, is that it has to evaluate every single record. So even if you see 5 rows being displayed, if you have 15x records bound to the `Data` parameter then it has to evaluate the 15 records to see which one's actually comply to the filter. So it, being called 15 times does not seem wrong, Am I missing something here?

As for it being called initially...
I don't see anything wrong with it, it has to figure out which data to display initially, and custom filter should be evaluated to provide proper filtered data.
Even if the user hasn't input a filter, you as a developer can have your own filter rules that should apply on the initial load for example. I don't think It's wrong for datagrid to apply filters once it's initialized.

To give you an example:
-  It could be a Datagrid on a page where a `QueryString` parameter would be evaluated on the `CustomFilter` altering the data initially displayed according to the `QueryString` parameter received.

**Comment by barruka** [comment]
David I can see your point with the `QueryString` (I didn't think about that, my bad) but on the other hand, you got it wrong with the rows number.

In the provided example of the doc we have 5 rows:
```
@code{
    private List<Employee> employeeList = new() { new() { FirstName = "David" }, new() { FirstName = "Mladen" }, new() { FirstName = "John" }, new() { FirstName = "Ana" }, new() { FirstName = "Jessica" } };
}
```
And the event `CustomFilter` fires 15 times, so it fires 3 times x n rows.
If I change the code adding one more Employee (having a total of 6 items in the list) then `CustomFilter`  fires 18 times (3 x n rows).

**Comment by David-Moreira** [comment]
Yes, `QueryString `was just an example, there can be other use cases.

So with the other issuee, if that is the case then you're right. I will test that.

**Comment by barruka** [comment]
I'm very sorry, but I just can disagree with you. 

I've made a blank new blazor project: 'https://github.com/barruka/BlazoriseDataGrid'

```
dotnet new blazorserver -n BlazoriseDataGrid
```

Added packages:
```
  <ItemGroup>
    <PackageReference Include="Blazorise.Bootstrap" Version="1.0.2" />
    <PackageReference Include="Blazorise.DataGrid" Version="1.0.2" />
    <PackageReference Include="Blazorise.Icons.FontAwesome" Version="1.0.2" />
  </ItemGroup>
```

Added in '_Imports.razor':
```
@using Blazorise
```

Added at top of 'Program.cs':
```
using Blazorise;
using Blazorise.Bootstrap;
using Blazorise.Icons.FontAwesome;
```

In 'Program.cs', before 'var app = builder.Build();' added:
```
builder.Services
    .AddBlazorise(options =>
    {
        options.Immediate = true;
    })
    .AddBootstrapProviders()
    .AddFontAwesomeIcons();

```

In '/Pages/_Layout.cshtml' before closing '</head>':
```
    <!-- inside of head section -->
	<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/css/bootstrap.min.css" integrity="sha384-zCbKRCUGaJDkqS1kPbPd7TveP5iyJE0EjAuZQTgFLD2ylzuqKfdKlfG/eSrtxUkn" crossorigin="anonymous">
	<link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.15.4/css/all.css">

	<link href="_content/Blazorise/blazorise.css" rel="stylesheet" />
	<link href="_content/Blazorise.Bootstrap/blazorise.bootstrap.css" rel="stylesheet" />
```

In '/Pages/_Layout.cshtml' in body section after '@RenderBody()':
```
    <!-- inside of body section and after the div/app tag  -->
	<script src="https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.slim.min.js" integrity="sha384-DfXdz2htPH0lsSSs5nCTpuj/zy4C+OGpamoFVy38MVBnE+IbbVYUew+OrCXaRkfj" crossorigin="anonymous"></script>
	<script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js" integrity="sha384-9/reFTGAW83EW2RDu2S0VKaIzap3H66lZH81PoYlFhbGU+6BZp6G7niu735Sk7lN" crossorigin="anonymous"></script>
	<script src="https://cdn.jsdelivr.net/npm/bootstrap@4.6.1/dist/js/bootstrap.min.js" integrity="sha384-VHvPCCyXqtD5DqJeNxl2dtTyhF78xXNXdkwX1CZeRusQfRKp+tA7hAShOK/B/fQ2" crossorigin="anonymous"></script>
```

Run the project and hit '/datagridtest'. Check debug console:
```
OnCustomFilter 202203310853453928 David 
OnCustomFilter 202203310853453962 MLaden 
OnCustomFilter 202203310853453962 John 
OnCustomFilter 202203310853453962 Ana 
OnCustomFilter 202203310853453962 Jessica 
OnCustomFilter 202203310853455377 David 
OnCustomFilter 202203310853455378 MLaden 
OnCustomFilter 202203310853455378 John 
OnCustomFilter 202203310853455378 Ana 
OnCustomFilter 202203310853455378 Jessica 
OnCustomFilter 202203310853455991 David 
OnCustomFilter 202203310853455991 MLaden 
OnCustomFilter 202203310853455992 John 
OnCustomFilter 202203310853455992 Ana 
OnCustomFilter 202203310853455992 Jessica
```

**Comment by stsrki** [comment]
@barruka I have copied your example and I can also see only two calls, not three.

```
blazor.webassembly.js:1 OnCustomFilter 202203311026288310 David 
blazor.webassembly.js:1 OnCustomFilter 202203311026288320 MLaden 
blazor.webassembly.js:1 OnCustomFilter 202203311026288330 John 
blazor.webassembly.js:1 OnCustomFilter 202203311026288340 Ana 
blazor.webassembly.js:1 OnCustomFilter 202203311026288340 Jessica 
blazor.webassembly.js:1 OnCustomFilter 202203311026288920 David 
blazor.webassembly.js:1 OnCustomFilter 202203311026288930 MLaden 
blazor.webassembly.js:1 OnCustomFilter 202203311026288930 John 
blazor.webassembly.js:1 OnCustomFilter 202203311026288940 Ana 
blazor.webassembly.js:1 OnCustomFilter 202203311026288940 Jessica 
```

**Comment by stsrki** [comment]
fixed with #3653
